pkg/controller/machine-set-boot-image/helpers.go

```diff
@@ -0,0 +1,108 @@
+		return fmt.Errorf("MachineSet object was nil")
+	}
+	if providerSpec == nil {
+		return fmt.Errorf("ProviderSpec object was nil")
+	}
+	rawBytes, err := json.Marshal(providerSpec)
+	if err != nil {
+		return fmt.Errorf("marshal into machineset failed: %w", err)
+	}
+	ms.Spec.Template.Spec.ProviderSpec.Value = &kruntime.RawExtension{Raw: rawBytes}
+	return nil
+}
+
+// This function unmarshals the golden stream configmap into a coreos
+// stream object. Returns an error if the unmarshal fails.
+func unmarshalStreamDataConfigMap(cm *corev1.ConfigMap, st interface{}) error {
+	if err := json.Unmarshal([]byte(cm.Data[StreamConfigMapKey]), &st); err != nil {
+		return fmt.Errorf("failed to parse CoreOS stream metadata: %w", err)
+	}
+	return nil
+}
+
+// This function checks if an array of machineManagers contains the target apigroup/resource and returns
+// a bool(success/fail), a label selector to filter the target resource and an error, if any.
+func getMachineResourceSelectorFromMachineManagers(machineManagers []opv1.MachineManager, apiGroup opv1.MachineManagerMachineSetsAPIGroupType, resource opv1.MachineManagerMachineSetsResourceType) (bool, labels.Selector, error) {
+	// If no machine managers exist; exit the enqueue process without errors.
+	if len(machineManagers) == 0 {
+		klog.Infof("No machine manager were found, so no MAPI machinesets will be enqueued.")
+		return false, labels.Nothing(), nil
+	}
+	for _, machineManager := range machineManagers {
+		if machineManager.APIGroup == apiGroup && machineManager.Resource == resource {
+			if machineManager.Selection.Mode == opv1.Partial {
+				selector, err := metav1.LabelSelectorAsSelector(machineManager.Selection.Partial.MachineResourceSelector)
+				return true, selector, err
+			} else if machineManager.Selection.Mode == opv1.All {
+				return true, labels.Everything(), nil
+			}
+		}
+	}
+	return false, labels.Nothing(), nil
+}
+
+// Returns architecture type for a given machineset
+func getArchFromMachineSet(machineset *machinev1beta1.MachineSet) (arch string, err error) {
+
+	// Valid set of machineset/node architectures
+	validArchSet := sets.New[string]("arm64", "s390x", "amd64", "ppc64le")
+	// Check if the annotation enclosing arch label is present on this machineset
+	archLabel, archLabelMatch := machineset.Annotations[MachineSetArchAnnotationKey]
+	if archLabelMatch {
+		// Grab arch value from the annotation and check if it is valid
+		_, archLabelValue, archLabelValueFound := strings.Cut(archLabel, ArchLabelKey)
+		if archLabelValueFound && validArchSet.Has(archLabelValue) {
+			return archtranslater.RpmArch(archLabelValue), nil
+		}
+		return "", fmt.Errorf("invalid architecture value found in annotation: %s ", archLabel)
+	}
+	// If no arch annotation was found on the machineset, default to the control plane arch.
+	// return the architecture of the node running this pod, which will always be a control plane node.
+	klog.Infof("Defaulting to control plane architecture")
```

Comment:
question: how often would we get here? Is the log message for debugging? Not sure how much this would help

Reply:
I have not seen it happen in my testing, so it was mostly just a fail safe.